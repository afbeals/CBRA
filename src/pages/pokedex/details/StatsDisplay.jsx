// Extenal
import React from 'react';
import PropTypes from 'prop-types';

// Local

// Constants
const classname = 'stats';
const places = { 1: 'one', 10: 'tens', 100: 'hunds', 1000: 'thous' };
// Functions
const getLargestPlace = (input, dic) => {
  const getNumber = div => {
    if (div === 1) {
      return [dic[div], input];
    }
    const amt = Math.floor((input / div) % 10);
    const val = Math.floor((input / +`${div}0`) * 100);
    return (amt && [dic[div], val]) || getNumber(div / 10);
  };
  return getNumber(1000);
};

// Component
const StatsDisplay = ({ pClassname, stats }) => (
  <div data-testid="details-stats" className={`${pClassname}__${classname}`}>
    <ul className={`${pClassname}__${classname}__list`}>
      {stats.map(({ base_stat: amt, stat: { name: statName } }, index) => {
        const [place, placeAmt] = getLargestPlace(amt, places);
        return (
          <li key={`${index}-${statName}-${amt}`} className="stat">
            <ul className="stat__graph">
              {[...Array(50)].map((bar, barInx) => (
                <li
                  key={`${bar}-${barInx}`}
                  className={`${place} stat__graph__item`}
                >
                  {bar}
                </li>
              ))}
              <li
                data-testid="details-stats-indicator"
                style={{ height: `${placeAmt}%` }}
                className="stat__graph__background"
              />
            </ul>
            <div className="stat__name">{statName}</div>
          </li>
        );
      })}
    </ul>
  </div>
);

StatsDisplay.propTypes = {
  pClassname: PropTypes.string.isRequired,
  stats: PropTypes.array,
};

StatsDisplay.defaultProps = {
  stats: [],
};

export default StatsDisplay;

export { StatsDisplay };
