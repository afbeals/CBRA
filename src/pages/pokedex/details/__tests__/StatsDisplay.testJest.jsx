// External
import React from 'react';
import { render, cleanup } from '@testing-library/react';

// Local
import StatsDisplay from '../StatsDisplay';

// tests
describe('StatsDisplay', () => {
  afterEach(cleanup);
  it('should match StatsDisplay to snapshot', () => {
    const { getByTestId } = render(<StatsDisplay pClassname={'parent'} />);
    const statsDisplay = getByTestId('details-stats');
    expect(statsDisplay).toMatchSnapshot();
  });

  it('should have a background height of 40%', () => {
    const statsArray = [{ base_stat: 40, stat: { name: 'attack' } }];
    const { getByTestId } = render(
      <StatsDisplay pClassname={'parent'} stats={statsArray} />,
    );
    const statsDisplayBg = getByTestId('details-stats-indicator');
    expect(statsDisplayBg.style.height).toEqual('40%');
  });

  it('should have a background height of 70%', () => {
    const statsArray = [{ base_stat: 70, stat: { name: 'defense' } }];
    const { getByTestId } = render(
      <StatsDisplay pClassname={'parent'} stats={statsArray} />,
    );
    const statsDisplayBg = getByTestId('details-stats-indicator');
    expect(statsDisplayBg.style.height).toEqual('70%');
  });

  it('should have a stat title', () => {
    const statsArray = [{ base_stat: 40, stat: { name: 'attack' } }];
    const { getByText } = render(
      <StatsDisplay pClassname={'parent'} stats={statsArray} />,
    );
    const statsDisplay = getByText('attack');
    expect(statsDisplay).toBeTruthy();
  });
});
