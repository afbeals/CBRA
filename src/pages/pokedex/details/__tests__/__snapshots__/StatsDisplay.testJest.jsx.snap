// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatsDisplay should match StatsDisplay to snapshot 1`] = `
<div
  class="parent__stats"
  data-testid="details-stats"
>
  <ul
    class="parent__stats__list"
  />
</div>
`;
